client/list.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (C) 2021 VMware, Inc. All Rights Reserved.
+ *
+ * Licensed under the GNU Lesser General Public License v2.1 (the "License");
+ * you may not use this file except in compliance with the License. The terms
+ * of the License are located in the COPYING file of this distribution.
+ */
+
+/*
+ * Module   : list.c
+ *
+ * Abstract :
+ *
+ *            tdnfclientlib
+ *
+ *            client library
+ *
+ * Authors  : Nitesh Kumar (kunitesh@vmware.com)
+ */
+
+#include "includes.h"
+
+TDNF_ML_LIST* 
+TDNFMergeList(
+    TDNF_ML_LIST* listA,
+    TDNF_ML_LIST* listB
+    )
+{
+    uint32_t dwError = 0;
+    TDNF_ML_LIST* mergedList = NULL;
+
+    if ((listA == NULL) && (listB == NULL))
+    {
+        return NULL;
+    }
+
+    // Base cases
+    if (listA == NULL) 
+        return (listB);
+    else if (listB == NULL)
+        return (listA);
+
+    //Compare the URL Preference to sort the URL List in descending order. 
+    TDNF_ML_URL_INFO *urlA = listA->data;
+    TDNF_ML_URL_INFO *urlB = listB->data;
+
+    if ((urlA == NULL) || (urlB == NULL))
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (urlA->preference >= urlB->preference) 
+    {
+        mergedList = listA;
+        mergedList->next = TDNFMergeList(listA->next, listB);
+    }
+    else 
+    {
+        mergedList = listB;
+        mergedList->next = TDNFMergeList(listA, listB->next);
+    }
+
+cleanup:
+    return (mergedList);
+
+error:
```

Comment:
Whenever we use the `BAIL_ON_TDNF_ERROR` macro we need the `error:` label. The `cleanup` is just boilerplate, but since it's everywhere I don't think it's confusing.

I don't really like this coding style either, but we have it all over the code and we should be consistent.